packages/headless/src/controllers/recent-results-list/headless-recent-results-list.ts

```diff
@@ -0,0 +1,172 @@
+   */
+  options?: RecentResultsListOptions;
+}
+
+const defaultRecentResultsProps: Required<RecentResultsListProps> = {
+  initialState: {
+    results: [],
+  },
+  options: {
+    maxLength: 10,
+  },
+};
+
+export interface RecentResultsListInitialState {
+  /**
+   * The list of recent results.
+   * @defaultValue `[]`
+   */
+  results: Result[];
+}
+
+const initialStateSchema = new Schema<RecentResultsListInitialState>({
+  results: new ArrayValue({required: true}),
+});
+
+export interface RecentResultsListOptions {
+  /**
+   * The maximum number of results to retain in the list.
+   * @defaultValue `10`
+   */
+  maxLength: number;
+}
+
+const optionsSchema = new Schema<RecentResultsListOptions>({
+  maxLength: new NumberValue({required: true, min: 1}),
+});
+
+/**
+ * The `RecentResultsList` controller manages the user's recent results.
+ */
+export interface RecentResultsList extends Controller {
+  /**
+   * The state of the RecentResultsList controller.
+   * */
+  state: RecentResultsState;
+  /**
+   * Clears the recent results list.
+   */
+  clear(): void;
+  /**
+   * Adds result to the recent results list.
+   * @param index - The index of the clicked recent result in the list.
+   */
+  handleRecentResultClick(index: number): void;
+}
+
+export function validateRecentResultsProps(
+  engine: SearchEngine,
+  props?: RecentResultsListProps
+) {
+  validateOptions(
+    engine,
+    optionsSchema,
+    props?.options,
+    'buildRecentResultsList'
+  );
+  validateInitialState(
+    engine,
+    initialStateSchema,
+    props?.initialState,
+    'buildRecentResultsList'
+  );
+}
+
+/**
+ * Creates a `RecentResultsList` controller instance.
+ *
+ * @param engine - The headless engine.
+ * @param props - The configuration `RecentResultsList` properties.
+ * @returns A `RecentResultsList` controller instance.
+ * */
+export function buildRecentResultsList(
+  engine: SearchEngine,
+  props?: RecentResultsListProps
+): RecentResultsList {
+  if (!loadRecentResultsListReducer(engine)) {
+    throw loadReducerError;
+  }
+
+  const controller = buildController(engine);
+  const {dispatch} = engine;
+  const getState = () => engine.state;
+
+  const registrationProps: Required<RecentResultsListProps> = {
+    ...defaultRecentResultsProps,
+    ...props,
+  };
+
+  validateRecentResultsProps(engine, registrationProps);
+
+  const options = {
+    results: registrationProps.initialState.results,
+    maxLength: registrationProps.options.maxLength,
+  };
+
+  dispatch(registerRecentResults(options));
+
+  return {
+    ...controller,
+
+    get state() {
+      const state = getState();
+
+      return state.recentResults;
+    },
+
+    clear() {
+      dispatch(logClearRecentResults());
+      dispatch(clearRecentResults());
+    },
+
+    handleRecentResultClick(index: number) {
+      const errorMessage = new NumberValue({
+        required: true,
+        min: 0,
+        max: this.state.results.length,
+      }).validate(index);
```

Comment:
The other controllers do not validate method arguments, so I think it is okay to not validate here (the dispatched action would still have validation).

On the `max` specifically, I'm not sure if there is an off-by-one since we are expecting an index, not a length?
Personally, I would place the logic in the reducer instead of the action, to avoid the complexity that comes with writing and testing thunks.